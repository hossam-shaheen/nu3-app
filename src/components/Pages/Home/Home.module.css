.search-container {
  background-color: #fff;
  padding: 30px 15px;
  margin-bottom: 15px;
  box-shadow: 0 2px 4px 0 rgba(0, 0, 0, 0.2), 0 25px 50px 0 rgba(0, 0, 0, 0.1);
}

.result-count-container {
  background-color: #fff;
  padding: 15px;
}

.results-header {
  display: flex;
  justify-content: space-between;
  align-items: center;
  margin: 20px auto;
  padding-bottom: 15px;
  border-bottom: 1px dashed #ccc;
}
