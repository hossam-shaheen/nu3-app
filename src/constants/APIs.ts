export const SEARCH_BASE_URL = 'http://localhost:3000/data/search.json';
export const PRODUCT_BASE_URL = 'http://localhost:3000/data/product.json';