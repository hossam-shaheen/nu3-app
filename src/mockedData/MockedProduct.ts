export const mockedProduct = {
  product: {
    id: 4365311672401,
    title: "nu3 Bio Acerola",
    body_html:
      "nu3 Bio Acerola-Pulver: 100 g ✓ Natürliche Vitamin C-Quelle ✓ Ohne Zusatzstoffe wie Maltodextrin ✓ Schonend gefriergetrocknet. Jetzt im nu3 Shop bestellen!",
    vendor: "nu3",
    url: "https://www.nu3.de/products/nu3-bio-acerola-pulver",

    variants: [
      {
        id: 32511433474129,
        title: "150 g",
        price: "18.99",
        option1: "150 g",
      },
      {
        id: 32788214087761,
        title: "3 x 150 g",
        price: "51.99",
        option1: "3 x 150 g",
      },
    ],
    options: [
      {
        id: 5668092411985,
        values: ["150 g", "3 x 150 g"],
      },
    ],
    images: [
      {
        id: 15491638657105,
        src: "https://cdn.shopify.com/s/files/1/0095/2274/1329/products/7_252F9_252Fe_252F2_252F79e29bdc358cf46c6f6df40f78ec1cabeda1a36a_21032_Top.jpg?v=1630500085",
      },
      {
        id: 16206942535761,
        src: "https://cdn.shopify.com/s/files/1/0095/2274/1329/products/a_252F2_252F4_252Ff_252Fa24f85e82f207a80c119cb64b0fa99d3ac4def70_21452.jpg?v=1630500198",
      },
    ],
    image: {
      id: 15491638657105,
      src: "https://cdn.shopify.com/s/files/1/0095/2274/1329/products/7_252F9_252Fe_252F2_252F79e29bdc358cf46c6f6df40f78ec1cabeda1a36a_21032_Top.jpg?v=1630500085",
    },
  },
};
