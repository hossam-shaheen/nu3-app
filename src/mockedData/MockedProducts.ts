export const mockedProducts = {
  items: [
    {
      price: 50.99,
      image:
        "https://cdn.shopify.com/s/files/1/0095/2274/1329/products/8_252F1_252Fc_252Fc_252F81ccf47420c1f3a11717ff5ad7f93e3e13214556_21446_large.jpg?v=1620826303",
      vendor: "nu3",
      title: "nu3 Premium Skin & Hair Beauty",
      id: "4725777367121",
      link: "/products/nu3-premium-skin-and-hair-beauty",
    },
  ],
};
