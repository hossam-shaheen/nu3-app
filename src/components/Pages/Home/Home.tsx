import { FunctionComponent, useState } from 'react';
import { ErrorType, ProductsType } from 'src/interfaces/interfaces';
import Products from '../../Components/ProductsCards/Products/Products';
import SearchBar from '../../Components/Shared/SearchBar/SearchBar';
import ResultCount from '../../Components/Shared/ResultCount/ResultCount';
import Sort from '../../Components/Shared/Sort/Sort';
import { SEARCH_BASE_URL } from '../../../constants/APIs';
import classes from "./Home.module.css";

export const Home: FunctionComponent = (): JSX.Element => {

    const [products, setProducts] = useState<ProductsType>({
        items: [],
        totalItems: 0
    });
    const [loading, setLoading] = useState<boolean>(false);
    const [error, setError] = useState<ErrorType | null>(null);
    const [noResults, setNoResults] = useState<string>("");

    const getSearchResult = (searchKeyWord: string, productResults: ProductsType): void => {
        const filteredProducts = productResults.items.filter(product => product.title.toLowerCase().includes(searchKeyWord.toLowerCase()));
        const totalFilteredItems = filteredProducts.length;
        if (totalFilteredItems > 0) {
            setProducts({
                items: filteredProducts,
                totalItems: totalFilteredItems
            });
        }
    }

    const validateResults = (productResults: ProductsType): void => {
        if (productResults.totalItems > 0) {
            setNoResults("");
        } else {
            setNoResults("No results found");
        }
    }

    const onSearch = async (searchKeyWord: string) => {
        setError(null);
        if (searchKeyWord.length > 1) {
            setLoading(true);
            try {
                const response = await fetch(`${SEARCH_BASE_URL}`);
                const data = await response.json();
                validateResults(data.items)
                getSearchResult(searchKeyWord, data);
            } catch (error) {
                setError({
                    icon: "fas fa-exclamation-triangle",
                    message: "Failed to fetch data"
                });
            }
            setLoading(false);
        } else {
            setProducts({
                items: [],
                totalItems: 0
            });
            setNoResults("");
        }
    }

    const onSort = (sortKeyWord: string): void => {
        const sortedProducts = products.items.sort((a, b) => {
            switch (sortKeyWord) {
                case "Alphabetical":
                    return a.title.localeCompare(b.title);
                case "Low":
                    return a.price - b.price;
                case "High":
                    return b.price - a.price;
                case "Random":
                    return Math.random() - 0.5;
                default:
                    return 0;
            }
        });
        setProducts({
            items: sortedProducts,
            totalItems: products.totalItems
        });
    }

    return <>
        <div className={classes["search-container"]}>
            <SearchBar onSearch={onSearch} />
        </div>
        {products.totalItems > 0 && <div className={classes["result-count-container"]}>
            <div className={classes["results-header"]}>
                <ResultCount count={products.totalItems} />
                <Sort onSort={onSort} />
            </div>
            <Products products={products.items} />
        </div>}

        {(error && !loading && !noResults) && <div className='error-message'><i className={error?.icon}></i>{error?.message}</div>}
        {(!error && loading && !noResults) && <div className='loading'>Loading...</div>}
        {(!error && !loading && noResults) && <div className='no-results'>{noResults}</div>}
    </>
}
export default Home;
