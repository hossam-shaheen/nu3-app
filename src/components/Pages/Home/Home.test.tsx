import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { BrowserRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import { setupServer } from "msw/node";
import { rest } from "msw";
import { mockedProducts } from 'src/mockedData/MockedProducts';
import { SEARCH_BASE_URL } from 'src/constants/APIs';
import Home from './Home';


const server = setupServer(
    rest.get(`${SEARCH_BASE_URL}`, (req, res, ctx) => {
        return res(ctx.status(200), ctx.json(mockedProducts));
    }));

beforeAll(() => server.listen());
afterEach(() => server.resetHandlers());
afterAll(() => server.close());
test('Home type search and products should be displayed : Renders', () => {
    render(<Home />);

    const searchInput = screen.getByPlaceholderText("Search for a product");
    expect(searchInput).toBeInTheDocument();
});

test('Home type in search and 1 result should be displayed in the document: Renders', async () => {
    render(<BrowserRouter>
        <Home />
    </BrowserRouter>);


    const searchInput = screen.getByPlaceholderText("Search for a product");
    expect(searchInput).toBeInTheDocument();

    userEvent.type(searchInput, 'nu3');
    const searchInputValue = searchInput.getAttribute('value');
    expect(searchInputValue).toBe('nu3');

    const productCards = await screen.findAllByRole("listitem");
    expect(productCards.length).toBe(1);

    const productResult = await screen.findByText("1 result");
    expect(productResult).toBeInTheDocument();

    const sortSelect = await screen.findByTestId("sort-select");
    expect(sortSelect).toBeInTheDocument();
});


test('Home : Snapshot', () => {
    const component = renderer.create(<Home />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});