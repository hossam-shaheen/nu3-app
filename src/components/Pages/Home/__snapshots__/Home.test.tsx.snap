// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home : Snapshot 1`] = `
<div
  className="search-container"
>
  <div
    className="search-bar"
  >
    <input
      onChange={[Function]}
      placeholder="Search for a product"
      type="text"
      value=""
    />
  </div>
</div>
`;
